Ensure that non-EPSG:3857 vector tiles layers can be rendered in the correct place

nyalldawson opened the pull request:
This fixes the rendering of vector tiles layers which are constructed in any non web Mercator CRS.

It also ensures that vector tiles loaded from an ESRI VectorTileService correctly read and initialise the tile structure, so that these layers will be correctly placed.

(Unfortunately the required information is not recorded in mbtiles packages, so it's necessary to set the correct parameters via
API calls when trying to load non-3857 mbtiles.)

Fixes https://github.com/qgis/QGIS/issues/39061

nyalldawson committed:
Ensure that non-EPSG:3857 vector tiles layers can be rendered in the
correct place

This fixes the rendering of vector tiles layers which are constructed
in any non web mercator CRS.

(Unfortunately the required information is not recorded in mbtiles
packages, so it's necessary to set the correct parameters via
API calls when trying to load non-3857 mbtiles.)

nyalldawson committed:
Correctly load arcgis vector tiles in non-EPSG:3857 crs

nyalldawson committed:
Spelling

nirvn commented:
Very, very good fix.

nirvn commented on src/core/vectortile/qgsvectortilestructure.cpp:
```diff
@@ -0,0 +1,90 @@
+{
+  QgsVectorTileStructure res;
+  res.setCrs( QgsCoordinateReferenceSystem( QStringLiteral( "EPSG:3857" ) ) );
+  res.setZ0xMin( -20037508.3427892 );
```
Do we have an established convention on setZ0xMin vs setZ0xMinimum?

nyalldawson replied:
You're right, fixed

wonder-sk commented on src/core/vectortile/qgsvectortilestructure.cpp:
```diff
@@ -0,0 +1,90 @@
+#include "qgsvectortileutils.h"
+#include "qgsarcgisrestutils.h"
+
+QgsVectorTileStructure QgsVectorTileStructure::fromWebMercator()
```
I believe with custom CRS the approach with bounds() is not going to be great. In some countries they have kind of standard tile matrix set for their national CRS, but I doubt it would be the same as the bounds() most of the time (just because of the fact that you want tile matrix boundary to be a square or built by multiple squares). It would be more appropriate to have a collection of "well known tile matrix sets" mainly for countries where there's such agreement (I don't know of any authority for this, but I have seen that MapTiler have a list of such tile matrix sets for some European countries).

nyalldawson commented on src/core/vectortile/qgsvectortilestructure.h:
```diff
@@ -0,0 +1,222 @@
+ * \ingroup core
+ * \since QGIS 3.22.6
+ */
+class CORE_EXPORT QgsVectorTileStructure
```
@wonder-sk 

>  think all of it is useful also for raster XYZ tiles and WMTS.

That was my original thought too. But the z0xmin/max etc used for scaling the vector tiles IS vt specific. And that's a large part of this class...

nyalldawson replied:
@wonder-sk 

Ok, I've refactored this to use QgsTileMatrixSet, with a subclass of QgsVectorTileMatrixSet which adds just the vector tile specific bits on top.

nyalldawson replied:
@wonder-sk 

To further clarify -- I've taken this as far as I'm able to justify timewise, but there's still some oddness in QgsTileMatrix (outside of this PR, that is). QgsTileMatrix has members for matrixWidth and matrixHeight, which I'd expect to contain the sizes of the matrix at a particular zoom level -- but these are always the root zoom 0 widths/heights (and also aren't used anywhere outside of tests!). Similarly, I'd expect QgsTileMatrix::extent to return the extent of the corresponding zoom level, but again it's always returning zoom level 0's extent.

So we're now closer to proper handling of individual zoom levels, but not quite there yet.

nyalldawson replied:
Ping @wonder-sk -- are you happy if I merge the new version?

nyalldawson committed:
Switch to using QgsMatrixTileSet as common tile set class

nyalldawson committed:
Fix build

nyalldawson committed:
Add tests

nyalldawson committed:
Restore strange scale doubling hack

wonder-sk commented on src/core/vectortile/qgsvectortilematrixset.cpp:
```diff
@@ -0,0 +1,104 @@
+    const int level = lodMap.value( QStringLiteral( "level" ) ).toInt();
+    if ( level == 0 )
+    {
+      z0Dimension = lodMap.value( QStringLiteral( "resolution" ) ).toDouble() * 512;
```
Not sure, but maybe if tileInfo.cols (or rows) could be used instead of hardcoded 512?

nyalldawson replied:
Right -- that looks correct! (All the services I can find have a row/col size of 512 anyway)

wonder-sk commented on src/core/vectortile/qgsvectortilematrixset.h:
```diff
@@ -0,0 +1,130 @@
+    double mZ0xMin = 0;
+    double mZ0xMax = 0;
+    double mZ0yMin = 0;
+    double mZ0yMax = 0;
```
This is where I am confused - why is this necessary to explicitly keep the Z0 rectangle here (rather than accessing extent() from zoom level 0) - is there any difference?

If we need to keep this explicitly, why not use QgsRectangle?

nyalldawson replied:
Right -- I can confirm that these always exactly match the results from tileMatrix(0).extent(). I've dropped them.

nyalldawson committed:
Address some review comments

nyalldawson committed:
Take row size from json instead of hardcoded value

nyalldawson committed:
Remove redunant z0 extent members

nyalldawson committed:
Don't apply tile scale doubling hack to esri generated vector tiles

We shouldn't apply this hack if we want to match ESRI's zoom to scale
level handling
